### ISSUE TITLE
Termination prover not strong enough

### ISSUE THREAD

**Issue by acjf3** [HOL-Theorem-Prover/HOL#101]
The following fails:

val test_def = Define`
   test n l s = if n = 0 then l else test (n - 1) (s::l) s`;

All variants fail as well. I actually want something similar to

val test_def = Define`
   test (l, n) s = if n = 0 then l else test (s::l, n - 1) s`;

or

val test_def = Define`
   test (n, l) s = if n = 0 then l else test (n - 1, s::l) s`;
##

**Comment by xrchz** [comment]
@acjf3 did you receive anything from Konrad regarding this?
I observe that your original `test_def` now (1fffd4b) appears to work without requiring any manual termination proof.

**Comment by acjf3** [comment]
I'm not aware of a check-in to address this. Note that the definition of interest

val test_def = Define`test (l, n) s = if n = 0 then l else test (s::l, n - 1) s`;

still fails.

**Comment by konrad-slind** [comment]
Right, I haven't done anything on improving the termination prover, which
really
should get that. It may be interesting to note that currying the function
unpacks the "real" termination problem so that the terminator gets it:

 Define `test l n s = if n=0 then l else test (s::l) (n-1) s`;
<<HOL message: inventing new type variable names: 'a>>
Equations stored under "test_def".
Induction stored under "test_ind".
val it =
   |- ∀s n l. test l n s = if n = 0 then l else test (s::l) (n − 1) s: thm

Thinking about it, the problem is that the straightforward application of
the size measure treats the pair

   (s::l,n-1)

as increasing by 1 in the first arg and decreasing by 1 in the second arg.
A more
refined approach would be to follow that blunt instrument approach with a
pass that
tried to see if any projection from any argument tuple decreases in size.

Konrad.

**Comment by konrad-slind** [comment]
Looking at this a bit more, if I swap the order of arguments, Define gets
the termination proof:

 Define `test1 s (l,n) = if n=0 then l else test1 s (s::l,n-1)`;
<<HOL message: inventing new type variable names: 'a>>
Equations stored under "test1_def".
Induction stored under "test1_ind".
val it =
   |- ∀s n l. test1 s (l,n) = if n = 0 then l else test1 s (s::l,n − 1):
   thm

So this is definitely an implementation bug: argument order shouldn't
matter.

(Notes following are possibly understandable to KLS only)
Culprit seems to be TotalDefn.guessR, which in the original only returns the
two candidate WF relations:

   guessR defn;
val it =
   [``inv_image $< (λ(x,y). 0)``,
    ``inv_image $< (λ(x,v). list_size (λv. 0) (FST x) + SND x)``]:
   term list

Neither of which work. Swapping argument order lets guessR do the
right thing:

 guessR defn1;
val it =
   [``inv_image $< (λ(x,y). 0)``,
    ``inv_image $< (λ(v,y). list_size (λv. 0) (FST y) + SND y)``,
    ``inv_image $< (λ(v1,v2,v3). v3)``]:
   term list

The last measure works.

Note in TotalDefn about how guessR works:

(*---------------------------------------------------------------------------*)
(* "guessR" guesses a list of termination measures. Quite ad hoc.
 *)
(* First guess covers recursions on proper subterms, e.g. prim. recs. Next
  *)
(* guess measure sum of sizes of all arguments. Next guess generates
  *)
(* lex-combos for Ackermann-style iterated prim. rec. defs. Finally,
  *)
(* generate all lex combinations of arguments that get changed by known
 *)
(* functions, i.e., ones that have corresponding theorems in ref variable
 *)
(* "termination_simps".
 *)
(* Finally, all generated termination relations are simplified and
  *)
(* duplicates are weeded out.
 *)
(*---------------------------------------------------------------------------*)
